flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java

```diff
@@ -81,14 +92,36 @@ NettyPartitionRequestClient createPartitionRequestClient(ConnectionID connection
 				Object old = clients.putIfAbsent(connectionId, connectingChannel);
 
 				if (old == null) {
-					nettyClient.connect(connectionId.getAddress()).addListener(connectingChannel);
-
-					client = connectingChannel.waitForChannel();
-
-					clients.replace(connectionId, connectingChannel, client);
+					int count = 0;
+					ChannelException exception = null;
+					do {
+						try {
+							exception = null;
+							connectingChannel.error = null;
+							LOG.info("Connecting to {} at {} attempt", connectionId.getAddress(), count);
+							nettyClient.connect(connectionId.getAddress()).addListener(connectingChannel);
+
+							client = connectingChannel.waitForChannel();
+
+							clients.replace(connectionId, connectingChannel, client);
+						} catch (ChannelException e) {
+							LOG.error("Failed {} times to connect to {}", count, connectionId.getAddress(), e);
+							exception = e;
+						}
+						count++;
+					} while (count <= retryNumber);
+
+					if (exception != null) {
+						throw exception;
+					}
 				}
 				else if (old instanceof ConnectingChannel) {
-					client = ((ConnectingChannel) old).waitForChannel();
+
+					try {
+						client = ((ConnectingChannel) old).waitForChannel();
+					} catch (IOException e) {
+						LOG.info("Continue to waitForChannel until the original thread has completed or failed");
+					}
```

Comment:
What do you mean by this change/log message?

Reply:
As multiple subpartition will share the same channel. Let's say subpartition a, b both needs the channel. A goes to wait first, B will still wait here. If exception happens, both of them will be notified. But we don't want the exception throw in B as A can still retry.

Reply:
But shouldn't `B` implement the same retry logic? Otherwise we are swallowing exceptions.

Reply:
The current logic seems have some problems. When it exists exception while waiting, then the client is still `null` which is replaced into the map via below `clients.replace(connectionId, old, client)`. In next loop, it would enter the above `connectChannelWithRetry` logic, that means the `B` might also encounter the same process as previous `A` to retry some times.

I guess the expectation we want is only let `A` retry connections and `B` is always waiting during `A` retries until final success or failure.

Reply:
@pnowojski 
The logic is always to let the first requester to retry. For the following requesters, will always go to wait until the first requester get the client or finished retry. Otherwise, there will be more retry times than expected.

@zhijiangW 
I am not sure whether my understanding is correct or now. Once the ConnectingChannel is put into clients concurrentMap by the first requester, the object is never null after that. Thus, other requesters will go to line 97. So the process is what you described " let A retry connections and B is always waiting during A retries until final success or failure." Are you aligned?

Reply:
Ok. I see, I missed the outer `while (client == null)` loop. 

But as it is now, isn't it possible that `B`  will loop indefinitely? Shouldn't it also follow the max retry count logic and if exceeded, throw an exception?

Reply:
Agree. Apply the same retry logic for following channel request other than the first one.